rfc/0001-shapely-2.0-roadmap.md

```diff
@@ -0,0 +1,441 @@
+>>> print(mp)
+MULTIPOINT (1 1, 2 2, 3 3)
+
+# using list(..) as way to iterate over it
+>>> list(p)
+...
+TypeError: 'Point' object is not iterable
+>>> list(mp)
+[<shapely.geometry.point.Point at 0x7f98475a0d90>,
+ <shapely.geometry.point.Point at 0x7f98475a0550>,
+ <shapely.geometry.point.Point at 0x7f98475a0d30>]
+# getitem also works
+>>> mp[0]
+<shapely.geometry.point.Point at 0x7f98462cab50>
+```
+
+This can certainly be convenient in certain cases, but also creates an
+inconsistency between the different geometry types and it creates some problems
+in other cases. For example, putting such scalars in a numpy object array:
+
+```python
+# single point is fine
+>>> np.array([p], dtype=object)
+array([<shapely.geometry.point.Point object at 0x7f98476cb2e0>],
+      dtype=object)
+
+# MultiPoint raises an error
+>>> np.array([mp], dtype=object)
+...
+TypeError: object of type 'Point' has no len()
+
+# MultiPolygon gets splitted
+>>> from shapely.geometry import MultiPolygon, box
+>>> mpoly = MultiPolygon([box(0, 0, 1, 1), box(2, 2, 3, 3)])
+>>> np.array([mpoly], dtype=object)
+array([[<shapely.geometry.polygon.Polygon object at 0x7f9846379ac0>,
+        <shapely.geometry.polygon.Polygon object at 0x7f9846379670>]],
+      dtype=object)
+```
+
+When working with arrays of geometry objects, the "iterability" of the
+geometries becomes an annoyance.
+
+The "Multi" geometry types also have a `.geoms` attribute, which gives access to
+a list of geometries to iterate over. So the question here can be: isn't this
+sufficient? Accessing this specific attribute doesn't seem problematic, as you
+need to know anyway that you have a "Multi" geometry, as the others will raise an
+error on iter/getitem.
+
+**Proposal for Shapely 2.0** Deprecate and eventually remove the iterability of
+"Multi" geometries, and point users to the `.geoms` attribute.
+
+This means that code like `mpoly[0]` or `for geom in mpoly: ...` needs to be
+replaced with `mpoly.geoms[0]` or `for geom in mpoly.geoms: ...`.
+
+
+### Empty geometries
+
+TO FILL IN
+
+Shapely 1.x is inconsistent in treating empty geometries, cfr https://github.com/Toblerity/Shapely/issues/742.
+
+
+### Spatial indexing with STRTree
+
+TO FILL IN
+
+PyGEOS also implements an STRtree. However, the API is not compatible with the
+one that is now in Shapely (but more similar to the API of `rtree`): the tree in
+PyGEOS returns integer indices, while the tree of Shapely returns actual
+geometry objects (cfr https://github.com/Toblerity/Shapely/issues/615).
+
+The behaviour of Shapely can easily be implemented on top of the behaviour of
+PyGEOS (to avoid a duplicate tree implementation). But the question here will
+still be how we want to provide this in the public API. Can we somehow unify
+this? Or just provide two tree objects users can choose from? Or is it
+acceptable to deprecate the current shapely behaviour? (or provide a keyword to
+still return geometries instead of integers?)
+
```

Comment:
How about for 2.0 we replace strtree.STRTree with a class that stores only integer indexes of the items in the sequence we initialize from. Shapely's object-storing tree was an attempt to make it very user friendly, but in the end it's not useful for geopandas, which already has its own object storage.

```
class STRTree:
    """Create an STR-packed R-tree from a sequence of geoms"""
    def intersection(geom):
        """Return indexes of tree leaves that intersect with geom"""
    def nearest(geom):
        """Return the index of the tree leaf nearest geom"""
```

Reply:
It's pretty easy to just `iloc` in GeoPandas